GPT Review development-guidelines (#280)
crytic/building-secure-contracts: development-guidelines/token_integration.md

# Token Integration Checklist

> This checklist offers recommendations for engaging with arbitrary tokens. Ensure that every unchecked item is justified and that its risks are understood.

Reviewer: ```suggestion
This checklist offers recommendations for interacting with arbitrary tokens. Ensure that every unchecked item is justified and that its risks are understood.
```

- slither-prop . --contract ContractName # requires configuration, and use of Echidna and Manticore
```

## General Considerations

> - [ ] **The contract has a security review.** Avoid engaging with contracts that lack a security review. Assess the review's duration (i.e., the level of effort), the reputation of the security firm, and the number and severity of findings.

Reviewer: ```suggestion
- [ ] **The contract has a security review.** Avoid interacting with contracts that lack a security review. Assess the review's duration (i.e., the level of effort), the reputation of the security firm, and the number and severity of findings.
```

- [ ] **The contract uses `SafeMath`.** Contracts that do not use `SafeMath` require a higher standard of review. Inspect the contract manually for `SafeMath` usage.
- [ ] **The contract has only a few non-token-related functions.** Non-token-related functions increase the likelihood of issues in the contract. Use Slither’s [`contract-summary` printer](https://github.com/crytic/slither/wiki/Printer-documentation#contract-summary) to broadly review the code used in the contract.
- [ ] **The token has only one address.** Tokens with multiple entry points for balance updates can break internal bookkeeping based on the address (e.g., `balances[token_address][msg.sender]` might not reflect the actual balance).

## Owner Privileges

- [ ] **The token is not upgradeable.** Upgradeable contracts may change their rules over time. Use Slither’s [`human-summary` printer](https://github.com/crytic/slither/wiki/Printer-documentation#contract-summary) to determine if the contract is upgradeable.

> - [ ] **The owner has limited minting capabilities.** Malicious or compromised owners can abuse minting capabilities. Use Slither’s [`human-summary` printer](https://github.com/crytic/slither/wiki/Printer-documentation#contract-summary) to review minting capabilities and consider manual code review.

Reviewer: ```suggestion
- [ ] **The owner has limited minting capabilities.** Malicious or compromised owners can abuse minting capabilities. Use Slither’s [`human-summary` printer](https://github.com/crytic/slither/wiki/Printer-documentation#contract-summary) to review minting capabilities and consider manually reviewing the code.
```

- [ ] **The `name`, `decimals`, and `symbol` functions are present if used.** These functions are optional in the ERC721 standard and may not be present.
- [ ] **If it is used, `decimals` returns a `uint8(0)`.** Other values are invalid.
- [ ] **If used, `decimals` returns a `uint8(0)`.** Other values are invalid.
- [ ] **The `name` and `symbol` functions can return an empty string.** This behavior is allowed by the standard.
- [ ] **The `ownerOf` function reverts if the `tokenId` is invalid or is set to a token that has already been burned.** The function cannot return 0x0. This behavior is required by the standard, but it is not always properly implemented.
- [ ] **The `ownerOf` function reverts if the `tokenId` is invalid or refers to a token that has already been burned.** The function cannot return 0x0. This behavior is required by the standard but may not always be implemented correctly.
- [ ] **A transfer of an NFT clears its approvals.** This is required by the standard.

> - [ ] **The token ID of an NFT cannot be changed during its lifetime.** This is required by the standard.

Reviewer: Revert deletion if unintentional
